src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs

```diff
@@ -250,6 +262,83 @@ private static IntPtr GetSslMethod(bool isServer, long options)
             return method;
         }
 
+        private static int verify_callback(int preverify_ok, IntPtr x509_ctx_ptr)
+        {
+            using (SafeX509StoreCtxHandle storeHandle = new SafeX509StoreCtxHandle(x509_ctx_ptr, false))
+            {
+                using (var chain = new X509Chain())
+                {
+                    chain.ChainPolicy.RevocationMode = X509RevocationMode.Online;
+                    chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+
+                    using (SafeX509StackHandle chainStack = libcrypto.X509_STORE_CTX_get1_chain(storeHandle))
+                    {
+                        if (chainStack.IsInvalid)
+                        {
+                            Debug.Fail("Invalid chain stack handle");
+                            return 0;
+                        }
+
+                        IntPtr certPtr = Crypto.GetX509StackField(chainStack, 0);
+                        if (IntPtr.Zero == certPtr)
+                        {
+                            return 0;
+                        }
+
+                        using (X509Certificate2 cert = new X509Certificate2(certPtr))
+                        {
+                            return chain.Build(cert) ? 1 : 0;
+                        }
+                    }
+                }
+            }
+        }
+
+        private static void UpdateCAListFromRootStore(libssl.SafeSslContextHandle context)
+        {
+            using (SafeX509NameStackHandle nameStack = Crypto.NewX509NameStack())
+            {
+                //Enumerate Certificates from LocalMachine and CurrentUser root store 
+                AddX509Name(nameStack, StoreLocation.LocalMachine);
+                AddX509Name(nameStack, StoreLocation.CurrentUser);
+                libssl.SSL_CTX_set_client_CA_list(context, nameStack);
+
+                // The handle ownership has been transferred into the CTX.
+                nameStack.SetHandleAsInvalid();
+            }
+
+        }
+
+        private static void AddX509Name(SafeX509NameStackHandle nameStack, StoreLocation storeLocation)
+        {
+            using (var myStore = new X509Store(StoreName.Root, storeLocation))
+            {
+                myStore.Open(OpenFlags.ReadOnly);
+
+                foreach (var cert in myStore.Certificates)
+                {
+                    using (SafeX509Handle certHandle = new SafeX509Handle(cert.Handle, false))
```

Comment:
Personally, I'd just call X509_dup instead of adding the ctor to specify `ownsHandle: false`.

In this case, since no one else has a copy of the certificates collection from this store object there's no chance of a parallel dispose race condition; but the cost of X509_dup is low compared to unsafe patterns.